core/src/impl/Kokkos_BitOps.hpp

```diff
@@ -57,93 +57,193 @@ namespace Kokkos {
 namespace Impl {
 
 KOKKOS_FORCEINLINE_FUNCTION
-int int_log2(unsigned i) {
+int int_log2_fallback(unsigned i) {
   enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
-#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+  int offset = 0;
+  if (i) {
+    for (offset = shift; (i & (1 << offset)) == 0; --offset)
+      ;
+  }
+  return offset;
+}
+
+KOKKOS_IMPL_DEVICE_FUNCTION
+inline int int_log2_device(unsigned i) {
+  enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return shift - __clz(i);
-#elif defined(KOKKOS_COMPILER_INTEL)
+#else
+  return int_log2_fallback(i);
+#endif
+}
+
+KOKKOS_IMPL_HOST_FUNCTION
+inline int int_log2_host(unsigned i) {
+  enum : int { shift = sizeof(unsigned) * CHAR_BIT - 1 };
+#if defined(KOKKOS_COMPILER_INTEL)
   return _bit_scan_reverse(i);
 #elif defined(KOKKOS_COMPILER_CRAYC)
   return i ? shift - _leadz32(i) : 0;
 #elif defined(__GNUC__) || defined(__GNUG__)
   return shift - __builtin_clz(i);
 #else
-  int offset = 0;
-  if (i) {
-    for (offset = shift; (i & (1 << offset)) == 0; --offset)
-      ;
-  }
-  return offset;
+  return int_log2_fallback(i);
+#endif
+}
+
+#if defined(__EDG__)
+#pragma push
+#pragma diag_suppress implicit_return_from_non_void_function
 #endif
+KOKKOS_FORCEINLINE_FUNCTION
+int int_log2(unsigned i) {
+  KOKKOS_IF_ON_DEVICE((return int_log2_device(i);))
+  KOKKOS_IF_ON_HOST((return int_log2_host(i);))
 }
+#if defined(__EDG__)
+#pragma pop
+#endif
 
 /**\brief  Find first zero bit.
  *
  *  If none then return -1 ;
  */
 KOKKOS_FORCEINLINE_FUNCTION
-int bit_first_zero(unsigned i) noexcept {
+int bit_first_zero_fallback(unsigned i) noexcept {
   enum : unsigned { full = ~0u };
+  int offset = -1;
+  if (full != i) {
+    for (offset = 0; i & (1 << offset); ++offset)
+      ;
+  }
+  return offset;
+}
 
-#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
+KOKKOS_IMPL_DEVICE_FUNCTION
+inline int bit_first_zero_device(unsigned i) noexcept {
+  enum : unsigned { full = ~0u };
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
   return full != i ? __ffs(~i) - 1 : -1;
-#elif defined(KOKKOS_COMPILER_INTEL)
+#else
+  return bit_first_zero_fallback(i);
```

Comment:
```suggestion
#elif defined(KOKKOS_COMPILER_INTEL)
  return full != i ? _bit_scan_forward(~i) : -1;
#else
  return bit_first_zero_fallback(i);
```
... and here as well.